core/shared/src/main/scala/zio/logging/LogFilter.scala

```diff
@@ -0,0 +1,375 @@
+   *
+   * @param rootLevel Default log level
+   * @param group Log group
+   * @param matcher Mather for log group and groupings
+   * @param groupings Log levels definitions
+   * @return A filter for log filtering based on given groups
+   */
+  def logLevelByGroup[A](
+    rootLevel: LogLevel,
+    group: LogGroup[A],
+    matcher: (A, A) => Boolean,
+    groupings: (A, LogLevel)*
+  ): LogFilter[Any] =
+    (trace, _, level, _, _, context, _, annotations) => {
+      val loggerGroup = group(trace, level, context, annotations)
+
+      val groupingLogLevel = groupings.collectFirst {
+        case (groupingGroup, groupingLevel) if matcher(loggerGroup, groupingGroup) => groupingLevel
+      }.getOrElse(rootLevel)
+
+      level >= groupingLogLevel
+    }
+
+  /**
+   * Defines a filter from a list of log-levels specified per tree node
+   *
+   * Example:
+   *
+   * {{{
+   *   val filter =
+   *     logLevelAndName(
+   *      LogLevel.Debug,
+   *      "io.netty"                                       -> LogLevel.Info,
+   *      "io.grpc.netty"                                  -> LogLevel.Info
+   * )
+   * }}}
+   *
+   * will use the `Debug` log level for everything except for log events with the logger name
+   * prefixed by either `List("io", "netty")` or `List("io", "grpc", "netty")`.
+   * Logger name is extracted from [[Trace]].
+   *
+   * @param rootLevel Minimum log level for the root node
+   * @param mappings  List of mappings, nesting defined by dot-separated strings
+   * @return A filter for log filtering based on log level and name
+   */
+  def logLevelByName(rootLevel: LogLevel, mappings: (String, LogLevel)*): LogFilter[Any] =
+    logLevelByGroup(rootLevel, LogGroup.loggerName, mappings: _*)
+
+  /**
+   * Defines a filter from a list of log-levels specified per tree node
+   *
+   * Example:
+   *
+   * {{{
+   *   val filter =
+   *     logLevelByGroup(
+   *      LogLevel.Debug,
+   *      "io.netty"                                       -> LogLevel.Info,
+   *      "io.grpc.netty"                                  -> LogLevel.Info
+   * )
+   * }}}
+   *
+   * will use the `Debug` log level for everything except for log events with the logger name
+   * prefixed by either `List("io", "netty")` or `List("io", "grpc", "netty")`.
+   *
+   * @param rootLevel  Minimum log level for the root node
+   * @param group Log group
+   * @param mappings   List of mappings, nesting defined by dot-separated strings
+   * @return A filter for log filtering based on log level and name
+   */
+  def logLevelByGroup(
+    rootLevel: LogLevel,
+    group: LogGroup[String],
+    mappings: (String, LogLevel)*
+  ): LogFilter[Any] = {
+    val mappingsSorted = mappings.map(splitNameByDotAndLevel.tupled).sorted(nameLevelOrdering)
+    val nameGroup      = group.map(splitNameByDot)
+    val matcher        = (loggerName: List[String], groupName: List[String]) => loggerName.startsWith(groupName)
```

Comment:
This example is pretty compelling for why `matcher` should be inside `LogGroup`: because anytime you construct `LogGroup[String]` based on log name, you probably want startsWith equivalence.